File: leetcode-project-2025-06-18T20-08-20/src/heap.ts
```typescript
// src/heap.ts
class MinHeap {
  heap: number[];

  constructor() {
    this.heap = [];
  }

  // ... (Implementation of heapifyUp, heapifyDown, insert, extractMin, etc.)
  insert(value: number): void {
    this.heap.push(value);
    this.heapifyUp(this.heap.length - 1);
  }


  extractMin(): number | null {
    if (this.heap.length === 0) return null;
    if (this.heap.length === 1) return this.heap.pop();

    const min = this.heap[0];
    this.heap[0] = this.heap.pop()!; // replace with last element
    this.heapifyDown(0);
    return min;
  }

  private heapifyUp(index: number): void {
    // ... (Implementation)
  }
  private heapifyDown(index: number): void {
    // ... (Implementation)
  }


  // Add peek, size, isEmpty, etc., as needed.
}


// Example usage and problem solutions would go here.  For instance, a solution to "Find Kth Largest Element" might look like this:

function findKthLargest(nums: number[], k: number): number {
  const minHeap = new MinHeap();
  for (const num of nums) {
    minHeap.insert(num);
    if (minHeap.heap.length > k) {
      minHeap.extractMin();
    }
  }
  return minHeap.heap[0]; //Min-heap will have the kth largest at the top
}



// ... (Implement other problem solutions here: Heap Sort, Merge K Sorted Lists, Top K Frequent Elements)

export { MinHeap, findKthLargest }; //Export relevant functions

```

File: leetcode-heap-operations-solutions-plus/heap.cpp
#include "heap.h"

template <typename T>
MinHeap<T>::MinHeap(int capacity) : capacity(capacity), heap(capacity) {}

template <typename T>
void MinHeap<T>::insert(T value) {
    if (size() == capacity) {
        std::cerr << "Heap is full!" << std::endl;
        return;
    }
    heap.push_back(value);
    heapifyUp(size() - 1);
}

template <typename T>
T MinHeap<T>::extractMin() {
    if (isEmpty()) {
        std::cerr << "Heap is empty!" << std::endl;
        return T{}; // Return default value for the type T
    }
    T minVal = heap[0];
    heap[0] = heap.back();
    heap.pop_back();
    heapifyDown(0);
    return minVal;
}


// ... (Implementation of other MinHeap methods, MaxHeap, and helper functions) ...

//Example for heapifyUp
template <typename T>
void MinHeap<T>::heapifyUp(int index) {
    int parent = (index - 1) / 2;
    while (index > 0 && heap[index] < heap[parent]) {
        std::swap(heap[index], heap[parent]);
        index = parent;
        parent = (index - 1) / 2;
    }
}

template <typename T>
bool MinHeap<T>::isEmpty() { return heap.empty(); }
template <typename T>
int MinHeap<T>::size() { return heap.size(); }
template <typename T>
T MinHeap<T>::peekMin() { return heap[0]; }


//Similar implementations for MaxHeap would be added here.